From 4a89f73d680b1fb2dd4862a6b9437628be9ad43f Mon Sep 17 00:00:00 2001
From: Daniel DeGrasse <daniel.degrasse@nxp.com>
Date: Wed, 24 Aug 2022 18:38:14 -0500
Subject: [PATCH] tests: boot: add mcuboot test

Add test to verify mcuboot support. This test is only enabled for specific
platforms, since it it not possible to filter for mcuboot support.
Sysbuild support is required to flash multiple samples using twister.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
---
 tests/boot/test_mcuboot/CMakeLists.txt        |  9 ++++++
 tests/boot/test_mcuboot/README.rst            | 12 ++++++++
 tests/boot/test_mcuboot/prj.conf              |  6 ++++
 tests/boot/test_mcuboot/src/main.c            | 19 ++++++++++++
 .../test_mcuboot/swapped_app/CMakeLists.txt   |  9 ++++++
 tests/boot/test_mcuboot/swapped_app/Kconfig   |  9 ++++++
 tests/boot/test_mcuboot/swapped_app/prj.conf  |  4 +++
 .../boot/test_mcuboot/swapped_app/src/main.c  | 13 ++++++++
 tests/boot/test_mcuboot/sysbuild.cmake        | 30 +++++++++++++++++++
 tests/boot/test_mcuboot/sysbuild.conf         |  1 +
 tests/boot/test_mcuboot/testcase.yaml         | 15 ++++++++++
 11 files changed, 127 insertions(+)
 create mode 100644 tests/boot/test_mcuboot/CMakeLists.txt
 create mode 100644 tests/boot/test_mcuboot/README.rst
 create mode 100644 tests/boot/test_mcuboot/prj.conf
 create mode 100644 tests/boot/test_mcuboot/src/main.c
 create mode 100644 tests/boot/test_mcuboot/swapped_app/CMakeLists.txt
 create mode 100644 tests/boot/test_mcuboot/swapped_app/Kconfig
 create mode 100644 tests/boot/test_mcuboot/swapped_app/prj.conf
 create mode 100644 tests/boot/test_mcuboot/swapped_app/src/main.c
 create mode 100644 tests/boot/test_mcuboot/sysbuild.cmake
 create mode 100644 tests/boot/test_mcuboot/sysbuild.conf
 create mode 100644 tests/boot/test_mcuboot/testcase.yaml

diff --git a/tests/boot/test_mcuboot/CMakeLists.txt b/tests/boot/test_mcuboot/CMakeLists.txt
new file mode 100644
index 0000000000..421792db4d
--- /dev/null
+++ b/tests/boot/test_mcuboot/CMakeLists.txt
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(test_mcuboot)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
diff --git a/tests/boot/test_mcuboot/README.rst b/tests/boot/test_mcuboot/README.rst
new file mode 100644
index 0000000000..60c5608088
--- /dev/null
+++ b/tests/boot/test_mcuboot/README.rst
@@ -0,0 +1,12 @@
+MCUBoot Swap Testing
+#####################
+
+Tests MCUBoot's image swap support. This application is built in three parts
+using sysbuild. The first application is the MCUBoot bootloader. The second
+application is the main sysbuild target, and will request an image swap
+from MCUBoot when booted. The third application is build with a load address
+adjustment using CONFIG_BUILD_OUTPUT_ADJUST_LMA, and will be the application
+that MCUBoot swaps to when the image swap is requested.
+
+This sequence of applications allows the test to verify support for the MCUBoot
+upgrade process on any platform supporting it.
diff --git a/tests/boot/test_mcuboot/prj.conf b/tests/boot/test_mcuboot/prj.conf
new file mode 100644
index 0000000000..6e1932f186
--- /dev/null
+++ b/tests/boot/test_mcuboot/prj.conf
@@ -0,0 +1,6 @@
+# Enable zephyr dfu boot utility library
+CONFIG_IMG_MANAGER=y
+# Enable flash drivers for MCUBoot
+CONFIG_FLASH=y
+# Enable reboot API to reset board after selecting new image
+CONFIG_REBOOT=y
diff --git a/tests/boot/test_mcuboot/src/main.c b/tests/boot/test_mcuboot/src/main.c
new file mode 100644
index 0000000000..265788fcc7
--- /dev/null
+++ b/tests/boot/test_mcuboot/src/main.c
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2022 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/reboot.h>
+#include <zephyr/dfu/mcuboot.h>
+
+/* Main entry point */
+void main(void)
+{
+	printk("Launching primary slot application on %s\n", CONFIG_BOARD);
+	/* Perform a permanent swap of MCUBoot application */
+	boot_request_upgrade(1);
+	printk("Secondary application ready for swap, rebooting\n");
+	sys_reboot(SYS_REBOOT_COLD);
+}
diff --git a/tests/boot/test_mcuboot/swapped_app/CMakeLists.txt b/tests/boot/test_mcuboot/swapped_app/CMakeLists.txt
new file mode 100644
index 0000000000..421792db4d
--- /dev/null
+++ b/tests/boot/test_mcuboot/swapped_app/CMakeLists.txt
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(test_mcuboot)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
diff --git a/tests/boot/test_mcuboot/swapped_app/Kconfig b/tests/boot/test_mcuboot/swapped_app/Kconfig
new file mode 100644
index 0000000000..9f603e6ce3
--- /dev/null
+++ b/tests/boot/test_mcuboot/swapped_app/Kconfig
@@ -0,0 +1,9 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+source "Kconfig.zephyr"
+# Workaround for not being able to have commas in macro arguments
+DT_CHOSEN_Z_CODE_PARTITION := zephyr,code-partition
+
+config BUILD_OUTPUT_ADJUST_LMA
+	default "$(dt_node_reg_addr_hex,$(dt_nodelabel_path,slot1_partition))-\
+		$(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_CODE_PARTITION))"
diff --git a/tests/boot/test_mcuboot/swapped_app/prj.conf b/tests/boot/test_mcuboot/swapped_app/prj.conf
new file mode 100644
index 0000000000..b3b57b2477
--- /dev/null
+++ b/tests/boot/test_mcuboot/swapped_app/prj.conf
@@ -0,0 +1,4 @@
+# Since LMA adjustment is used to move the image location to the secondary
+# slot, we need to enable hex build output. Otherwise, some debuggers will not
+# flash the binary to the second image slot.
+CONFIG_BUILD_OUTPUT_HEX=y
diff --git a/tests/boot/test_mcuboot/swapped_app/src/main.c b/tests/boot/test_mcuboot/swapped_app/src/main.c
new file mode 100644
index 0000000000..f3b9030f4d
--- /dev/null
+++ b/tests/boot/test_mcuboot/swapped_app/src/main.c
@@ -0,0 +1,13 @@
+/*
+ * Copyright 2022 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+
+/* Main entry point */
+void main(void)
+{
+	printk("Swapped application booted on %s\n", CONFIG_BOARD);
+}
diff --git a/tests/boot/test_mcuboot/sysbuild.cmake b/tests/boot/test_mcuboot/sysbuild.cmake
new file mode 100644
index 0000000000..c3a467dbef
--- /dev/null
+++ b/tests/boot/test_mcuboot/sysbuild.cmake
@@ -0,0 +1,30 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+# Add the mcuboot key file to the secondary swapped app
+
+if(SB_CONFIG_BOOTLOADER_MCUBOOT)
+  set(swapped_app_CONFIG_BOOTLOADER_MCUBOOT y CACHE STRING
+      "MCUBOOT is enabled as bootloader" FORCE
+  )
+  set(swapped_app_CONFIG_MCUBOOT_SIGNATURE_KEY_FILE
+      \"${SB_CONFIG_BOOT_SIGNATURE_KEY_FILE}\" CACHE STRING
+      "Signature key file for signing" FORCE
+  )
+endif()
+
+# Add the swapped app to the build
+ExternalZephyrProject_Add(
+  APPLICATION swapped_app
+  SOURCE_DIR ${APP_DIR}/swapped_app
+)
+
+# Add the swapped app to the list of images to flash
+# Ensure the order of images is as follows:
+# - mcuboot
+# - swapped app
+# - primary app (test_mcuboot)
+# This order means that if the debugger resets the MCU in between flash
+# iterations, the MCUBoot swap won't be triggered until the secondary app
+# is actually present in flash.
+set(IMAGES "mcuboot" "swapped_app" "test_mcuboot")
diff --git a/tests/boot/test_mcuboot/sysbuild.conf b/tests/boot/test_mcuboot/sysbuild.conf
new file mode 100644
index 0000000000..47f00ff3cf
--- /dev/null
+++ b/tests/boot/test_mcuboot/sysbuild.conf
@@ -0,0 +1 @@
+SB_CONFIG_BOOTLOADER_MCUBOOT=y
diff --git a/tests/boot/test_mcuboot/testcase.yaml b/tests/boot/test_mcuboot/testcase.yaml
new file mode 100644
index 0000000000..37b2eabfa5
--- /dev/null
+++ b/tests/boot/test_mcuboot/testcase.yaml
@@ -0,0 +1,15 @@
+common:
+    sysbuild: True
+    harness: console
+    harness_config:
+      type: multi_line
+      regex:
+        - "I: Starting bootloader"
+        - "Launching primary slot application on (.*)"
+        - "Secondary application ready for swap, rebooting"
+        - "I: Starting swap using (.*)"
+        - "Swapped application booted on (.*)"
+tests:
+  boot.mcuboot:
+    tags: mcuboot
+    platform_allow: frdm_k64f mimxrt1060_evk
-- 
2.37.3

